### ISSUE TITLE
Migration to argparse from optparse

### ISSUE THREAD

**Issue by Pierre-Sassoulas** [PyCQA/pylint#5392]
### Current problem

``optparse`` is deprecated since python 3.2 and is not developed anymore. Also, the parsing are handled in a strange way with callback that created stupid bug that could probably be avoided by refactoring the code (https://github.com/PyCQA/pylint/pull/3581) We need to migrate to another CLI parser, the most logical being ``argparse`` or ``click``, but ``confuse`` is also a possibility because it permits to have default option in a configuration file that could be one of the configuration we already handle and make option / configuration consistent. I'm open to other libraries too. 

### Desired solution

No more optparse used in the code, small and modular handling of CLI argument parsing, no more special case, or minimal.

### Additional context

https://github.com/PyCQA/pylint/pull/5152#issuecomment-960890607
https://github.com/PyCQA/pylint/issues/5259#issuecomment-960879430

**Comment by DanielNoord** [comment]
I think `argparse` makes the most sense as it is the direct successor to `optparse`.

To do this I think it makes the most sense to do some initial preparation, namely convert all options to a new class (we could use the one I proposed in #5152) and move them all to `PyLinter`. This avoids the current problem we're having with double settings etc and will probably also be faster (?) as we only need to parse arguments once instead for each individual checker.

**Comment by saroad2** [comment]
Even though `argparse` is built-in library, and a way better solution than `optparse`, I **strongly recommend** using [Click](https://click.palletsprojects.com/en/8.0.x/) as your CLI library:

* It makes a shorter and cleaner code
* It has a lot of neat features that makes it so much easy to write CLI with.
* It is a living project that still publish new abilities and features
* It is supported by its developers
* It's widely used in the industry (personally, I use it all the time and in all of my projects)

**Comment by Pierre-Sassoulas** [comment]
I like click, but my concern is it adds a new dependency. Also [argparse is compatible with confuse](https://confuse.readthedocs.io/en/latest/usage.html#command-line-options) which would be very nice to help  with consistency between options and configuration file enforced in a simple way.

**Comment by saroad2** [comment]
In my humble opinion, adding a dependency is not that of a problem. In the end it's all matter of "buy what you won't implement", using a library instead of developing something out of your domain. I think that `click` answers to this definition.

As to `confuse`, I am not familiar with this library so I don't have a strong opinion about it. In the end it's all matter of pros and cons. From my own experience with both `argparse` and `click`, I think `click` brings much more wholistic solution to the CLI programming process. But again, one an argue otherwise.

**Comment by DanielNoord** [comment]
I have started work on migration to `argparse`. I have not previous experience with any of the alternatives mentioned in this issue, but I think trying to depend only on `stdlib` is a good thing for a static checker such as `pylint`. It allows everybody to use `pylint` in combination with other linter or analysers.

One of the things that I think we should change (and I'm changing in my proposal) is that option are no longer registered to each individual checker but to the `PyLinter` object. This is preparing for (eventually) supporting sub-directory configuration:
If we create helper functions that can be passed an `argparse.Namespace` (or look up the correct `Namespace`) and return the current value, we can create `Namespaces` for individual directory's. This is similar to how we're already using `get_global_option`. We would just need to enforce the use of these helpers more strictly, but I think that approach might work (and fix a long standing issue with `pylint`).

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I want to ask for input before committing to this:
Do we want to keep requiring the sections in options files? It adds a lot of extra checks in the code and honestly I don't see why we should require it. It's trivial to keep supporting it but I would like to support:
```toml
[tool.pylint]
missing-member-hint = true

[tool.pylint."TYPECHECK"]
missing-member-hint = true

[tool.pylint."I-ORDER-MY-SECTIONS-THE-WAY-I-WANT-TO"]
missing-member-hint = true
```

I do not immediately see what having these sections internally offers to us.

**Comment by Pierre-Sassoulas** [comment]
> I do not immediately see what having these sections internally offers to us.

Only upside would be to have the same option within two checkers with a different value in each checkers ? :trollface: Imo not an upside and more like a nightmare. Otherwise I don't know. I think we expose our internals checker name to users so it makes it more complex for us to change anything in the checkers and more complex for users to understand what they need to add in the configuration. See https://github.com/PyCQA/pylint/issues/3181, with currentely 37 upvotes from confused configuration writers. One of the top upvoted issues we have.

Imo, the only reason it stays this way is the enormous amount of existing config and doc about config in the wild. I agree with you that we should support all three possibilities, maybe not take the sections into account at all. This is the kind of thing a migration tool would fix / clean up / normalize to:
```
[tool.pylint]
missing-member-hint = true
```

**Comment by DanielNoord** [comment]
I'll go ahead with that implementation then! 😄

**Comment by DanielNoord** [comment]
With #6278 almost merged we're nearing the process of completion on this particular issue. After that there are still many more issues on the project that should be merged before releasing `2.14`, but I think we are nearing a state in which we could potentially do a pre-release to see if I haven't broken anything.

What's left to do:
- [ ] Stop parsing with `optparse` completely
- [ ] Use `config` instead of `namespace` again in all of our own code
- [ ] Remove `namespace` attribute in favour of `config` on `ArgumentsManager`

**Comment by Pierre-Sassoulas** [comment]
I think we should rename namespace to config as you said or we'll still be impacted by optparse in 2030. Ideally we would want a solution that do not need to have two branch depending on the pylint version in libraries depending on pylint's code but if it's not possible, well... 

We could raise an error explaining what needs to be done to lib mainteners instead of simply removing the function, I think it's better than outright getting a no-member without explanation (then we really remove it in 3.0).

**Comment by DanielNoord** [comment]
Do you want to go as far as changing `ArgumentsManager.load_command_line_configuration` (for example) to become an empty shell with a `DeprecationWarning` and a longer message explaining what anybody should be doing?
Or would you still want to keep those methods even though they might only partially work?

**Comment by DanielNoord** [comment]
> * [x]  Stop parsing with `optparse` completely
> * [x]  Use `config` instead of `namespace` again in all of our own code
> * [x]  Remove `namespace` attribute in favour of `config` on `ArgumentsManager`

With #6319 this will be done.

Two last tasks remain for this to be closed and a pre-release to be releasable:
- [ ] Fix any outstanding coverage issues in `pylint/config`
- [ ] Go through all `TODOs` and see if any need immediate fixing

**Comment by DanielNoord** [comment]
I'm going to leave the lower coverage in ``__init__`` and ``find_default_config_files.py`` as those are not related to the recent changes. I'd like to finish this as soon as possible.

Updating the TODO above.

**Comment by DanielNoord** [comment]
One final TODO list and we're done with this issue:

- [ ] #6334
- [ ] Make sure we don't use deprecated classes in our tests

**Comment by DanielNoord** [comment]
Going to close this as there is nothing left to do here 😄

**Comment by jacobtylerwalls** [comment]
A lot of useful changes came from this project, but I just wonder if @DanielNoord has any insight into the [proposal](https://github.com/python/cpython/issues/126180) to undeprecate `optparse`.

**Comment by DanielNoord** [comment]
Yeah I saw that: they should definitely undeprecate it!

I think the main benefit we got out of this that it forced me to redo most of our configuration handling which 1) gave me a lot of insight into the codebase and ideas to improve it and 2) forced us to refactor and standardize some undefined behaviour.
`argparse` itself hasn't been that useful compared to `optparse`. Both are really not that user friendly to work with and lead to weird behaviour (as described in the issue).

I'm glad that they deprecated it as it gave us a reason to have a good look at this code, but as to `argparse` being any better than `optparse` that would be a definitive +0 from me.
